File: backend/app/models/flight.py
import enum
from datetime import datetime

from sqlalchemy import String, DateTime, Enum
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.core.database import Base


class FlightStatus(str, enum.Enum):
    scheduled = "scheduled"
    confirmed = "confirmed"
    boarding = "boarding"
    departed = "departed"
    in_air = "in_air"
    landed = "landed"
    cancelled = "cancelled"
    delayed = "delayed"


class Flight(Base):
    __tablename__ = "flights"

    id: Mapped[int] = mapped_column(primary_key=True)
    airline: Mapped[str] = mapped_column(String(255))
    flight_number: Mapped[str] = mapped_column(String(50))
    departure_airport: Mapped[str] = mapped_column(String(100))
    arrival_airport: Mapped[str] = mapped_column(String(100))
    departure_datetime: Mapped[datetime] = mapped_column(DateTime)
    arrival_datetime: Mapped[datetime] = mapped_column(DateTime)
    gate: Mapped[str | None] = mapped_column(String(20), nullable=True)
    seat_number: Mapped[str | None] = mapped_column(String(20), nullable=True)
    status: Mapped[FlightStatus] = mapped_column(
        Enum(FlightStatus), default=FlightStatus.scheduled
    )
    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        server_default=func.now(), onupdate=func.now()
    )

File: mobile/lib/features/dashboard/data/models/flight.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'flight.freezed.dart';
part 'flight.g.dart';

@freezed
class Flight with _$Flight {
  const factory Flight({
    required int id,
    required String airline,
    @JsonKey(name: 'flight_number') required String flightNumber,
    @JsonKey(name: 'departure_airport') required String departureAirport,
    @JsonKey(name: 'arrival_airport') required String arrivalAirport,
    @JsonKey(name: 'departure_datetime') required DateTime departureDatetime,
    @JsonKey(name: 'arrival_datetime') required DateTime arrivalDatetime,
    String? gate,
    @JsonKey(name: 'seat_number') String? seatNumber,
    required String status,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
  }) = _Flight;

  factory Flight.fromJson(Map<String, dynamic> json) => _$FlightFromJson(json);
}
